engineering-decision-records/017-version-and-migrate-db.md

```diff
@@ -0,0 +1,29 @@
+---
+type: architecture
+status: proposed
+participants: abailly
+---
+
+# Version & Migrate Chain Database
+
+## Context
+
+* While Amaru is not yet used in production, there's already been a couple occurences where developers found themselves stuck with an unusable database following upgrade of Amaru binary, requiring scrapping the existing data and boostrapping afresh. While bootstrapping is relatively fast, resynchronising is a lengthy process which one would rather like to avoid
+* This behaviour will be unacceptable in production especially as Amaru strives to provide fast turnaround and low footprint, and we need to preserve existing users' data as much as we possible
+* _Migration_ is a common feature afforded by classical database applications whereby the system checks the consistency of its database at boot time and either offers to migrate the data in case of incompatibility or even migrate it automatically
+* Migration tools are rife in the RDBMS world (see [this page](https://wiki.postgresql.org/wiki/Change_management_tools_and_techniques) for a Postgres-centric list) and is typically handled by storing versions and migration "scripts" in the database itself to make it self-checkable
```

Comment:
About `and migration "scripts" in the database itself`, in my experience ([flyway](https://www.red-gate.com/products/flyway/), [sqlx](https://crates.io/crates/sqlx)), the scripts (whether pure SQL or code) lives in the executable not in the database, which is what you describe later with the `migrate-chain-db` command.